### DISCUSSION TITLE
MVTLayer + ArcLayer

### DISCUSSION THREAD

**Discussion by hyundotio** [visgl/deck.gl#5367] [Q&A]
Hello,

I am a noobie when it comes to Deck, so I apologize in advance if this question is outright stupid heh.

I'm trying to create a ArcLayer (as a sublayer) from Mapbox Tiles (MVTLayer). When I assign `getSourceColor` and `getTargetColor` it starts and ends at tile seams, not the actual beginning + end.  This lead me to believe that the object coordinates are localized to tiles. And when I logged the object, that seemed to be the case.

I tried searching and searching and the closest thing I've gotten to was a thread that mentions MVT Loader / Loader Options, and some parameters (modelMatrix, coordinateSystem, coordinateOrigin, clipBounds)... but I had no luck in getting the result I want (a single line that has getSourceColor -> getTargetColor,   rather than repeating gradient per tile).

Any help or pointers would be appreciated.

Thank you.

![image](https://user-images.githubusercontent.com/20031831/105425430-61381280-5c17-11eb-88ef-dca9cc036cc6.png)

**Comment by kylebarron** [comment]
Vector tiles are tiled, which means that every tile is loaded and rendered independently. In order to correctly render the ArcLayer, you'd need to find a way to find the actual start and end point of each arc. It would probably be considerably easier to do this without storing the data as vector tiles.

**Reply by hyundotio** [reply]
Thank you very much. I understand. I hope you don't mind me going off-topic-ish for a second... but the design choice I made for this was to communicate a starting + end point of a line. it can be a gradient, arrows, markers, whatever... it just needs to be a direction.

Would it be possible to use something like PathMarkerLayer or would I run into the same tiling issues?
...or would it be just easier to move away from MVT?

**Reply by kylebarron** [reply]
You could store some metadata in your MVT so that you know which points are at the start/end of the line, and then render a scatterplot layer separately with different colors depending on if the point is at the start/end, etc.

**Reply by hyundotio** [reply]
> You could store some metadata in your MVT so that you know which points are at the start/end of the line, and then render a scatterplot layer separately with different colors depending on if the point is at the start/end, etc.

So I actually do this already. There are two layers:  a line layer + Icon layer. The beginning + end points have different icons. However, it is the line itself (especially when zoomed in) that can get disorienting. Ie there's a line from CA to NY, but when the user is zoomed in Colorado for example, the user has a solid line without any direction. It'd be great to either have a gradient, arrow marker, or something to show direction.

I hope this makes sense.

**Reply by kylebarron** [reply]
You'd have to do something like store a unique index for each start-end pair, but it's not generally possible because you have no idea that start and end are in the viewport.

**Reply by Pessimistress** [reply]
You can also use an IconLayer to place an arrow at the midpoint of the line segment.

**Reply by hyundotio** [reply]
@Pessimistress, this means I'd have to calculate the midpoint using the 2 localized coordinates then figure out the angle/rotation too correct?

**Reply by Pessimistress** [reply]
The localized coordinates are already in cartesian space, so the midpoint will be `[(x0+x1)/2, (y0+y1)/2]` and the rotation will be `Math.atan2(y1-y0, x1-x0)`.

**Reply by hyundotio** [reply]
@Pessimistress Thank you.  Then the Math.atan2(y1-y0, x1-x0) would be that * 180 / Pi - correct?

(Math.atan2(y1-y0, x1-x0) * 180) / Pi

Cheers!
